File: tests/python/test_aroon.py
"""
Python binding tests for Aroon indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestAroon:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_aroon_partial_params(self, test_data):
        """Test Aroon with partial parameters (None values) - mirrors check_aroon_partial_params"""
        high = test_data['high']
        low = test_data['low']


        result = ta_indicators.aroon(high, low, 14)
        up, down = result
        assert len(up) == len(high)
        assert len(down) == len(low)

    def test_aroon_accuracy(self, test_data):
        """Test Aroon matches expected values from Rust tests - mirrors check_aroon_accuracy"""
        high = test_data['high']
        low = test_data['low']
        expected = EXPECTED_OUTPUTS['aroon']

        result = ta_indicators.aroon(
            high, low,
            length=expected['default_params']['length']
        )
        up, down = result

        assert len(up) == len(high)
        assert len(down) == len(low)


        assert_close(
            up[-5:],
            expected['last_5_up'],
            rtol=0.0,
            atol=1e-2,
            msg="Aroon up last 5 values mismatch"
        )
        assert_close(
            down[-5:],
            expected['last_5_down'],
            rtol=0.0,
            atol=1e-2,
            msg="Aroon down last 5 values mismatch"
        )





    def test_aroon_default_candles(self, test_data):
        """Test Aroon with default parameters - mirrors check_aroon_default_candles"""
        high = test_data['high']
        low = test_data['low']


        result = ta_indicators.aroon(high, low, 14)
        up, down = result
        assert len(up) == len(high)
        assert len(down) == len(low)

    def test_aroon_zero_length(self):
        """Test Aroon fails with zero length - mirrors check_aroon_zero_length"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 11.0])

        with pytest.raises(ValueError, match="Invalid length"):
            ta_indicators.aroon(high, low, length=0)

    def test_aroon_length_exceeds_data(self):
        """Test Aroon fails when length exceeds data length - mirrors check_aroon_length_exceeds_data"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 11.0])

        with pytest.raises(ValueError, match="Invalid length"):
            ta_indicators.aroon(high, low, length=14)

    def test_aroon_very_small_dataset(self):
        """Test Aroon fails with insufficient data - mirrors check_aroon_very_small_data_set"""
        high = np.array([100.0])
        low = np.array([99.5])

        with pytest.raises(ValueError, match="Invalid length|Not enough valid data"):
            ta_indicators.aroon(high, low, length=14)

    def test_aroon_empty_input(self):
        """Test Aroon fails with empty input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.aroon(empty, empty, length=14)

    def test_aroon_mismatched_lengths(self):
        """Test Aroon fails with mismatched input lengths"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0])

        with pytest.raises(ValueError, match="High/low length mismatch"):
            ta_indicators.aroon(high, low, length=2)

    def test_aroon_reinput(self, test_data):
        """Test Aroon applied with different parameters - mirrors check_aroon_reinput"""
        high = test_data['high']
        low = test_data['low']


        first_result = ta_indicators.aroon(high, low, length=14)
        first_up, first_down = first_result
        assert len(first_up) == len(high)
        assert len(first_down) == len(low)


        second_result = ta_indicators.aroon(high, low, length=5)
        second_up, second_down = second_result
        assert len(second_up) == len(high)
        assert len(second_down) == len(low)

    def test_aroon_nan_handling(self, test_data):
        """Test Aroon handles NaN values correctly - mirrors check_aroon_nan_handling"""
        high = test_data['high']
        low = test_data['low']

        result = ta_indicators.aroon(high, low, length=14)
        up, down = result
        assert len(up) == len(high)
        assert len(down) == len(low)


        if len(up) > 240:
            assert not np.any(np.isnan(up[240:])), "Found unexpected NaN in up after warmup period"
            assert not np.any(np.isnan(down[240:])), "Found unexpected NaN in down after warmup period"


        expected_warmup = 14
        assert np.all(np.isnan(up[:expected_warmup])), "Expected NaN in up warmup period"
        assert np.all(np.isnan(down[:expected_warmup])), "Expected NaN in down warmup period"

    def test_aroon_streaming(self, test_data):
        """Test Aroon streaming API - mirrors check_aroon_streaming"""
        high = test_data['high']
        low = test_data['low']
        length = 14


        batch_result = ta_indicators.aroon(high, low, length=length)
        batch_up, batch_down = batch_result


        stream = ta_indicators.AroonStream(length=length)
        stream_up = []
        stream_down = []

        for h, l in zip(high, low):
            result = stream.update(h, l)
            if result is not None:
                stream_up.append(result[0])
                stream_down.append(result[1])
            else:
                stream_up.append(np.nan)
                stream_down.append(np.nan)

        stream_up = np.array(stream_up)
        stream_down = np.array(stream_down)


        assert len(batch_up) == len(stream_up)
        assert len(batch_down) == len(stream_down)


        mask_up = ~(np.isnan(batch_up) | np.isnan(stream_up))
        mask_down = ~(np.isnan(batch_down) | np.isnan(stream_down))

        if np.any(mask_up):
            assert_close(
                batch_up[mask_up],
                stream_up[mask_up],
                rtol=0.0,
                atol=1e-8,
                msg="Aroon up streaming mismatch"
            )

        if np.any(mask_down):
            assert_close(
                batch_down[mask_down],
                stream_down[mask_down],
                rtol=0.0,
                atol=1e-8,
                msg="Aroon down streaming mismatch"
            )

    def test_aroon_batch(self, test_data):
        """Test Aroon batch processing - mirrors check_batch_default_row"""
        high = test_data['high']
        low = test_data['low']

        result = ta_indicators.aroon_batch(
            high, low,
            length_range=(14, 14, 0),
        )

        assert 'up' in result
        assert 'down' in result
        assert 'lengths' in result


        up = result['up']
        down = result['down']
        assert up.shape[0] == 1
        assert up.shape[1] == len(high)
        assert down.shape[0] == 1
        assert down.shape[1] == len(low)


        default_up = up[0]
        default_down = down[0]
        expected = EXPECTED_OUTPUTS['aroon']


        assert_close(
            default_up[-5:],
            expected['last_5_up'],
            rtol=1e-2,
            msg="Aroon batch default up row mismatch"
        )
        assert_close(
            default_down[-5:],
            expected['last_5_down'],
            rtol=1e-2,
            msg="Aroon batch default down row mismatch"
        )

    def test_aroon_all_nan_input(self):
        """Test Aroon with all NaN values"""
        all_nan = np.full(100, np.nan)


        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.aroon(all_nan, all_nan, length=14)

    def test_aroon_batch_multiple_lengths(self, test_data):
        """Test Aroon batch with multiple lengths"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]

        result = ta_indicators.aroon_batch(
            high, low,
            length_range=(10, 20, 5),
        )

        assert 'up' in result
        assert 'down' in result
        assert 'lengths' in result


        up = result['up']
        down = result['down']
        assert up.shape[0] == 3
        assert up.shape[1] == 100
        assert down.shape[0] == 3
        assert down.shape[1] == 100


        assert len(result['lengths']) == 3
        assert list(result['lengths']) == [10, 15, 20]


        for i, length in enumerate([10, 15, 20]):
            row_up = up[i]
            row_down = down[i]
            expected_warmup = length

            assert np.all(np.isnan(row_up[:expected_warmup])), f"Expected NaN in up warmup for length {length}"
            assert np.all(np.isnan(row_down[:expected_warmup])), f"Expected NaN in down warmup for length {length}"

            if expected_warmup < 100:
                assert not np.all(np.isnan(row_up[expected_warmup:])), f"Expected values in up after warmup for length {length}"
                assert not np.all(np.isnan(row_down[expected_warmup:])), f"Expected values in down after warmup for length {length}"

    def test_aroon_kernel_selection(self, test_data):
        """Test Aroon with different kernel selections"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]


        result_scalar = ta_indicators.aroon(high, low, length=14, kernel="scalar")
        scalar_up, scalar_down = result_scalar
        assert len(scalar_up) == 100
        assert len(scalar_down) == 100


        result_auto = ta_indicators.aroon(high, low, length=14)
        auto_up, auto_down = result_auto
        assert len(auto_up) == 100
        assert len(auto_down) == 100



        mask_up = ~(np.isnan(scalar_up) | np.isnan(auto_up))
        mask_down = ~(np.isnan(scalar_down) | np.isnan(auto_down))

        if np.any(mask_up):
            assert_close(
                scalar_up[mask_up],
                auto_up[mask_up],
                rtol=0.0,
                atol=1e-10,
                msg="Kernel up results should match"
            )

        if np.any(mask_down):
            assert_close(
                scalar_down[mask_down],
                auto_down[mask_down],
                rtol=0.0,
                atol=1e-10,
                msg="Kernel down results should match"
            )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_aroon.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('Aroon partial params', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);

    const result = wasm.aroon_js(high, low, 14);
    assert(result.up, 'Should have up array');
    assert(result.down, 'Should have down array');
    assert.strictEqual(result.up.length, high.length);
    assert.strictEqual(result.down.length, low.length);
});

test('Aroon accuracy', async () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const expected = EXPECTED_OUTPUTS.aroon;

    const result = wasm.aroon_js(
        high, low,
        expected.defaultParams.length || 14
    );

    assert.strictEqual(result.up.length, high.length);
    assert.strictEqual(result.down.length, low.length);


    const last5Up = Array.from(result.up.slice(-5));
    const last5Down = Array.from(result.down.slice(-5));

    assertArrayClose(
        last5Up,
        expected.last5Up,
        1e-2,
        "Aroon up last 5 values mismatch"
    );
    assertArrayClose(
        last5Down,
        expected.last5Down,
        1e-2,
        "Aroon down last 5 values mismatch"
    );




});

test('Aroon default candles', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);

    const result = wasm.aroon_js(high, low, 14);
    assert.strictEqual(result.up.length, high.length);
    assert.strictEqual(result.down.length, low.length);
});

test('Aroon zero length', () => {

    const high = new Float64Array([10.0, 11.0, 12.0]);
    const low = new Float64Array([9.0, 10.0, 11.0]);

    assert.throws(() => {
        wasm.aroon_js(high, low, 0);
    }, /Invalid length/);
});

test('Aroon length exceeds data', () => {

    const high = new Float64Array([10.0, 11.0, 12.0]);
    const low = new Float64Array([9.0, 10.0, 11.0]);

    assert.throws(() => {
        wasm.aroon_js(high, low, 14);
    }, /Invalid length/);
});

test('Aroon very small dataset', () => {

    const high = new Float64Array([100.0]);
    const low = new Float64Array([99.5]);

    assert.throws(() => {
        wasm.aroon_js(high, low, 14);
    }, /Invalid length|Not enough valid data/);
});

test('Aroon empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.aroon_js(empty, empty, 14);
    }, /Input data slice is empty/);
});

test('Aroon mismatched lengths', () => {

    const high = new Float64Array([10.0, 11.0, 12.0]);
    const low = new Float64Array([9.0, 10.0]);

    assert.throws(() => {
        wasm.aroon_js(high, low, 2);
    }, /Mismatch in high\/low slice length/);
});

test('Aroon reinput', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);


    const firstResult = wasm.aroon_js(high, low, 14);
    assert.strictEqual(firstResult.up.length, high.length);
    assert.strictEqual(firstResult.down.length, low.length);


    const secondResult = wasm.aroon_js(high, low, 5);
    assert.strictEqual(secondResult.up.length, high.length);
    assert.strictEqual(secondResult.down.length, low.length);
});

test('Aroon NaN handling', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);

    const result = wasm.aroon_js(high, low, 14);
    assert.strictEqual(result.up.length, high.length);
    assert.strictEqual(result.down.length, low.length);


    if (result.up.length > 240) {
        for (let i = 240; i < result.up.length; i++) {
            assert(!isNaN(result.up[i]), `Found unexpected NaN in up at index ${i}`);
            assert(!isNaN(result.down[i]), `Found unexpected NaN in down at index ${i}`);
        }
    }


    assertAllNaN(Array.from(result.up.slice(0, 14)), "Expected NaN in up warmup period");
    assertAllNaN(Array.from(result.down.slice(0, 14)), "Expected NaN in down warmup period");
});

test('Aroon all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);


    assert.throws(() => {
        wasm.aroon_js(allNaN, allNaN, 14);
    }, /All values are NaN/, 'Expected AllValuesNaN error');
});

test('Aroon batch single parameter set', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);


    const batchResult = wasm.aroon_batch_js(
        high, low,
        14, 14, 0
    );


    const singleResult = wasm.aroon_js(high, low, 14);

    assert.strictEqual(batchResult.up.length, singleResult.up.length);
    assert.strictEqual(batchResult.down.length, singleResult.down.length);
    assertArrayClose(Array.from(batchResult.up), Array.from(singleResult.up), 1e-10, "Batch vs single up mismatch");
    assertArrayClose(Array.from(batchResult.down), Array.from(singleResult.down), 1e-10, "Batch vs single down mismatch");
});

test('Aroon batch multiple lengths', () => {

    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));


    const batchResult = wasm.aroon_batch_js(
        high, low,
        10, 20, 5
    );


    assert.strictEqual(batchResult.up.length, 3 * 100);
    assert.strictEqual(batchResult.down.length, 3 * 100);


    const lengths = [10, 15, 20];
    for (let i = 0; i < lengths.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowUp = batchResult.up.slice(rowStart, rowEnd);
        const rowDown = batchResult.down.slice(rowStart, rowEnd);

        const singleResult = wasm.aroon_js(high, low, lengths[i]);
        assertArrayClose(
            Array.from(rowUp),
            Array.from(singleResult.up),
            1e-10,
            `Length ${lengths[i]} up mismatch`
        );
        assertArrayClose(
            Array.from(rowDown),
            Array.from(singleResult.down),
            1e-10,
            `Length ${lengths[i]} down mismatch`
        );
    }
});

test('Aroon batch metadata', () => {

    const metadata = wasm.aroon_batch_metadata_js(
        10, 20, 5
    );


    assert.strictEqual(metadata.length, 3);


    assert.strictEqual(metadata[0], 10);
    assert.strictEqual(metadata[1], 15);
    assert.strictEqual(metadata[2], 20);
});

test('Aroon batch full parameter sweep', () => {

    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));

    const batchResult = wasm.aroon_batch_js(
        high, low,
        10, 14, 2
    );

    const metadata = wasm.aroon_batch_metadata_js(10, 14, 2);


    const numCombos = metadata.length;
    assert.strictEqual(numCombos, 3);
    assert.strictEqual(batchResult.up.length, 3 * 50);
    assert.strictEqual(batchResult.down.length, 3 * 50);


    for (let combo = 0; combo < numCombos; combo++) {
        const length = metadata[combo];
        const rowStart = combo * 50;
        const rowUp = Array.from(batchResult.up.slice(rowStart, rowStart + 50));
        const rowDown = Array.from(batchResult.down.slice(rowStart, rowStart + 50));


        for (let i = 0; i < Math.min(length, 50); i++) {
            assert(isNaN(rowUp[i]), `Expected NaN in up at warmup index ${i} for length ${length}`);
            assert(isNaN(rowDown[i]), `Expected NaN in down at warmup index ${i} for length ${length}`);
        }


        if (length < 50) {
            for (let i = length; i < 50; i++) {
                assert(!isNaN(rowUp[i]), `Unexpected NaN in up at index ${i} for length ${length}`);
                assert(!isNaN(rowDown[i]), `Unexpected NaN in down at index ${i} for length ${length}`);
            }
        }
    }
});

test('Aroon batch edge cases', () => {

    const high = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const low = new Float64Array([0.9, 1.9, 2.9, 3.9, 4.9, 5.9, 6.9, 7.9, 8.9, 9.9]);


    const singleBatch = wasm.aroon_batch_js(
        high, low,
        5, 5, 1
    );

    assert.strictEqual(singleBatch.up.length, 10);
    assert.strictEqual(singleBatch.down.length, 10);


    const largeBatch = wasm.aroon_batch_js(
        high, low,
        5, 7, 10
    );


    assert.strictEqual(largeBatch.up.length, 10);
    assert.strictEqual(largeBatch.down.length, 10);


    assert.throws(() => {
        wasm.aroon_batch_js(
            new Float64Array([]),
            new Float64Array([]),
            9, 9, 0
        );
    }, /All values are NaN|Empty input data|Invalid length|Not enough valid data/);
});


test('Aroon batch - new ergonomic API with single parameter', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);

    const result = wasm.aroon_batch(high, low, {
        length_range: [14, 14, 0]
    });


    assert(result.up, 'Should have up array');
    assert(result.down, 'Should have down array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, high.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.up.length, high.length);
    assert.strictEqual(result.down.length, low.length);


    const combo = result.combos[0];
    assert.strictEqual(combo.length, 14);


    const oldResult = wasm.aroon_js(high, low, 14);
    for (let i = 0; i < oldResult.up.length; i++) {
        if (isNaN(oldResult.up[i]) && isNaN(result.up[i])) {
            continue;
        }
        assert(Math.abs(oldResult.up[i] - result.up[i]) < 1e-10,
               `Up value mismatch at index ${i}`);
    }
    for (let i = 0; i < oldResult.down.length; i++) {
        if (isNaN(oldResult.down[i]) && isNaN(result.down[i])) {
            continue;
        }
        assert(Math.abs(oldResult.down[i] - result.down[i]) < 1e-10,
               `Down value mismatch at index ${i}`);
    }
});

test('Aroon batch - new API with multiple parameters', () => {

    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));

    const result = wasm.aroon_batch(high, low, {
        length_range: [10, 14, 2]
    });


    assert.strictEqual(result.rows, 3);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 3);
    assert.strictEqual(result.up.length, 150);
    assert.strictEqual(result.down.length, 150);


    const expectedLengths = [10, 12, 14];

    for (let i = 0; i < expectedLengths.length; i++) {
        assert.strictEqual(result.combos[i].length, expectedLengths[i]);
    }


    const secondRowUp = result.up.slice(result.cols, 2 * result.cols);
    const secondRowDown = result.down.slice(result.cols, 2 * result.cols);
    assert.strictEqual(secondRowUp.length, 50);
    assert.strictEqual(secondRowDown.length, 50);


    const oldResult = wasm.aroon_js(high, low, 10);
    const firstRowUp = result.up.slice(0, result.cols);
    const firstRowDown = result.down.slice(0, result.cols);
    for (let i = 0; i < oldResult.up.length; i++) {
        if (isNaN(oldResult.up[i]) && isNaN(firstRowUp[i])) {
            continue;
        }
        assert(Math.abs(oldResult.up[i] - firstRowUp[i]) < 1e-10,
               `Up value mismatch at index ${i}`);
    }
    for (let i = 0; i < oldResult.down.length; i++) {
        if (isNaN(oldResult.down[i]) && isNaN(firstRowDown[i])) {
            continue;
        }
        assert(Math.abs(oldResult.down[i] - firstRowDown[i]) < 1e-10,
               `Down value mismatch at index ${i}`);
    }
});

test('Aroon batch - new API matches old API results', () => {

    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));

    const params = {
        length_range: [10, 15, 5]
    };


    const oldResult = wasm.aroon_batch_js(
        high, low,
        params.length_range[0], params.length_range[1], params.length_range[2]
    );


    const newResult = wasm.aroon_batch(high, low, params);


    assert.strictEqual(oldResult.up.length, newResult.up.length);
    assert.strictEqual(oldResult.down.length, newResult.down.length);

    for (let i = 0; i < oldResult.up.length; i++) {
        if (isNaN(oldResult.up[i]) && isNaN(newResult.up[i])) {
            continue;
        }
        assert(Math.abs(oldResult.up[i] - newResult.up[i]) < 1e-10,
               `Up value mismatch at index ${i}: old=${oldResult.up[i]}, new=${newResult.up[i]}`);
    }
    for (let i = 0; i < oldResult.down.length; i++) {
        if (isNaN(oldResult.down[i]) && isNaN(newResult.down[i])) {
            continue;
        }
        assert(Math.abs(oldResult.down[i] - newResult.down[i]) < 1e-10,
               `Down value mismatch at index ${i}: old=${oldResult.down[i]}, new=${newResult.down[i]}`);
    }
});

test('Aroon batch - new API error handling', () => {
    const high = new Float64Array(testData.high.slice(0, 10));
    const low = new Float64Array(testData.low.slice(0, 10));


    assert.throws(() => {
        wasm.aroon_batch(high, low, {
            length_range: [9, 9]
        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.aroon_batch(high, low, {

        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.aroon_batch(high, low, {
            length_range: "invalid"
        });
    }, /Invalid config/);
});



test.after(() => {
    console.log('Aroon WASM tests completed');
});
